packages/optimism-decoder/src/decoder.ts

```diff
@@ -0,0 +1,261 @@
+import fs from "fs";
+import path from "path";
+import rlp from "rlp";
+
+import {
+  chunks,
+  bytesToNumber,
+  fetchAndExtractVersionedHashes,
+  downloadAndSaveFiles,
+  saveOptimismDecodedData,
+} from "./helpers";
+import { readVarint, readBytesAsHex, readBitlist } from "./reader";
+import { extractFullChannel, decompressBatches } from "./utils";
+
+export type OptimismDecodedData = {
+  timestampSinceL2Genesis: number;
+  lastL1OriginNumber: number;
+  parentL2BlockHash: string;
+  l1OriginBlockHash: string;
+  numberOfL2Blocks: number;
+  changedByL1Origin: number; // how many were changed by L1 origin
+  totalTxs: number;
+  contractCreationTxsNumber: number;
+  // TODO: add the following values
+  // legacyTxsNumber: number;
+  // totalGasLimit: number;
+  // protectedLegacyTxsNumber: number;
+};
+
+/**
+ * Main function to read and process the binary file.
+ * @param filename - The name of the file to read.
+ * This must be a file concatenating all blobs by frame number.
+ * @returns An object containing decoded fields.
+ */
+export async function decodeOptimismFile(
+  filename: string
+): Promise<OptimismDecodedData> {
+  const blobs = fs.readFileSync(filename);
+  const datas: Uint8Array[] = [];
+
+  for (const blob of chunks(blobs, 131072)) {
+    if (blob[1] !== 0) throw new Error("Assertion failed: blob[1] must be 0");
+    const declaredLength = bytesToNumber(blob.slice(2, 5));
+    let blobData = new Uint8Array();
+
+    for (const chunk of chunks(blob, 128)) {
+      // split into chunks of 128 bytes
+      const byteA = chunk[32 * 0];
+      const byteB = chunk[32 * 1];
+      const byteC = chunk[32 * 2];
+      const byteD = chunk[32 * 3];
+
+      if (
+        byteA === undefined ||
+        byteB === undefined ||
+        byteC === undefined ||
+        byteD === undefined
+      ) {
+        throw new Error("Assertion failed: bytes must be defined");
+      }
+
+      if ((byteA | byteB | byteC | byteD) & 0b1100_0000) {
+        throw new Error("Assertion failed: bytes must meet specific criteria");
+      }
+
+      const tailA = chunk.slice(32 * 0 + 1, 32 * 1);
+      const tailB = chunk.slice(32 * 1 + 1, 32 * 2);
+      const tailC = chunk.slice(32 * 2 + 1, 32 * 3);
+      const tailD = chunk.slice(32 * 3 + 1, 32 * 4);
+
+      const x = (byteA & 0b0011_1111) | ((byteB & 0b0011_0000) << 2);
+      const y = (byteB & 0b0000_1111) | ((byteD & 0b0000_1111) << 4);
+      const z = (byteC & 0b0011_1111) | ((byteD & 0b0011_0000) << 2);
+
+      const result = new Uint8Array(4 * 31 + 3);
```

Comment:
I would avoid magic numbers. We could create a const with a meaningful name. Also, it's repeated in `L85`)

Reply:
these numbers come from the specification